Multiple warnings when compiling datatable with macOS Clang 12.0.0

```
src/core/models/dt_linearmodel.cc:115:66: warning: implicit conversion from 'size_t' (aka 'unsigned long') to 'float' may lose precision [-Wimplicit-int-float-conversion]
  size_t last_iteration_nrows = static_cast<size_t>(last_epoch * iteration_nrows);
                                                               ~ ^~~~~~~~~~~~~~~
src/core/models/dt_linearmodel.cc:615:51: note: in instantiation of function template specialization 'dt::LinearModel<float>::fit_impl<signed char>' requested here
template LinearModelFitOutput LinearModel<float>::fit_impl<int8_t>();
                                                  ^
src/core/models/dt_linearmodel.cc:131:68: warning: implicit conversion from 'size_t' (aka 'unsigned long') to 'float' may lose precision [-Wimplicit-int-float-conversion]
    iteration_nrows = static_cast<size_t>(std::ceil(nepochs_val_ * iteration_nrows));
                                                                 ~ ^~~~~~~~~~~~~~~
src/core/models/dt_linearmodel.cc:227:47: warning: implicit conversion from 'size_t' (aka 'unsigned long') to 'float' may lose precision [-Wimplicit-int-float-conversion]
                betas_[i][j] += betas[i][j] / dt::num_threads_in_team();
                                            ~ ^~~~~~~~~~~~~~~~~~~~~~~~~
src/core/models/dt_linearmodel.cc:297:74: warning: implicit conversion from 'size_t' (aka 'unsigned long') to 'double' may lose precision [-Wimplicit-int-float-conversion]
  double epoch_stopped = static_cast<double>(iteration_end) / dt_X_fit_->nrows();
                                                            ~ ~~~~~~~~~~~^~~~~~~
src/core/models/dt_linearmodel.cc:115:66: warning: implicit conversion from 'size_t' (aka 'unsigned long') to 'float' may lose precision [-Wimplicit-int-float-conversion]
  size_t last_iteration_nrows = static_cast<size_t>(last_epoch * iteration_nrows);
                                                               ~ ^~~~~~~~~~~~~~~
src/core/models/dt_linearmodel.cc:616:51: note: in instantiation of function template specialization 'dt::LinearModel<float>::fit_impl<int>' requested here
template LinearModelFitOutput LinearModel<float>::fit_impl<int32_t>();
                                                  ^
src/core/models/dt_linearmodel.cc:131:68: warning: implicit conversion from 'size_t' (aka 'unsigned long') to 'float' may lose precision [-Wimplicit-int-float-conversion]
    iteration_nrows = static_cast<size_t>(std::ceil(nepochs_val_ * iteration_nrows));
                                                                 ~ ^~~~~~~~~~~~~~~
src/core/models/dt_linearmodel.cc:227:47: warning: implicit conversion from 'size_t' (aka 'unsigned long') to 'float' may lose precision [-Wimplicit-int-float-conversion]
                betas_[i][j] += betas[i][j] / dt::num_threads_in_team();
                                            ~ ^~~~~~~~~~~~~~~~~~~~~~~~~
src/core/models/dt_linearmodel.cc:297:74: warning: implicit conversion from 'size_t' (aka 'unsigned long') to 'double' may lose precision [-Wimplicit-int-float-conversion]
  double epoch_stopped = static_cast<double>(iteration_end) / dt_X_fit_->nrows();
                                                            ~ ~~~~~~~~~~~^~~~~~~
src/core/models/dt_linearmodel.cc:115:66: warning: implicit conversion from 'size_t' (aka 'unsigned long') to 'float' may lose precision [-Wimplicit-int-float-conversion]
  size_t last_iteration_nrows = static_cast<size_t>(last_epoch * iteration_nrows);
                                                               ~ ^~~~~~~~~~~~~~~
src/core/models/dt_linearmodel.cc:617:51: note: in instantiation of function template specialization 'dt::LinearModel<float>::fit_impl<float>' requested here
template LinearModelFitOutput LinearModel<float>::fit_impl<float>();
                                                  ^
src/core/models/dt_linearmodel.cc:131:68: warning: implicit conversion from 'size_t' (aka 'unsigned long') to 'float' may lose precision [-Wimplicit-int-float-conversion]
    iteration_nrows = static_cast<size_t>(std::ceil(nepochs_val_ * iteration_nrows));
                                                                 ~ ^~~~~~~~~~~~~~~
src/core/models/dt_linearmodel.cc:227:47: warning: implicit conversion from 'size_t' (aka 'unsigned long') to 'float' may lose precision [-Wimplicit-int-float-conversion]
                betas_[i][j] += betas[i][j] / dt::num_threads_in_team();
                                            ~ ^~~~~~~~~~~~~~~~~~~~~~~~~
src/core/models/dt_linearmodel.cc:297:74: warning: implicit conversion from 'size_t' (aka 'unsigned long') to 'double' may lose precision [-Wimplicit-int-float-conversion]
  double epoch_stopped = static_cast<double>(iteration_end) / dt_X_fit_->nrows();
                                                            ~ ~~~~~~~~~~~^~~~~~~
src/core/models/dt_linearmodel.cc:115:66: warning: implicit conversion from 'size_t' (aka 'unsigned long') to 'double' may lose precision [-Wimplicit-int-float-conversion]
  size_t last_iteration_nrows = static_cast<size_t>(last_epoch * iteration_nrows);
                                                               ~ ^~~~~~~~~~~~~~~
src/core/models/dt_linearmodel.cc:618:52: note: in instantiation of function template specialization 'dt::LinearModel<double>::fit_impl<signed char>' requested here
template LinearModelFitOutput LinearModel<double>::fit_impl<int8_t>();
                                                   ^
src/core/models/dt_linearmodel.cc:131:68: warning: implicit conversion from 'size_t' (aka 'unsigned long') to 'double' may lose precision [-Wimplicit-int-float-conversion]
    iteration_nrows = static_cast<size_t>(std::ceil(nepochs_val_ * iteration_nrows));
                                                                 ~ ^~~~~~~~~~~~~~~
src/core/models/dt_linearmodel.cc:227:47: warning: implicit conversion from 'size_t' (aka 'unsigned long') to 'double' may lose precision [-Wimplicit-int-float-conversion]
                betas_[i][j] += betas[i][j] / dt::num_threads_in_team();
                                            ~ ^~~~~~~~~~~~~~~~~~~~~~~~~
src/core/models/dt_linearmodel.cc:297:74: warning: implicit conversion from 'size_t' (aka 'unsigned long') to 'double' may lose precision [-Wimplicit-int-float-conversion]
  double epoch_stopped = static_cast<double>(iteration_end) / dt_X_fit_->nrows();
                                                            ~ ~~~~~~~~~~~^~~~~~~
src/core/models/dt_linearmodel.cc:115:66: warning: implicit conversion from 'size_t' (aka 'unsigned long') to 'double' may lose precision [-Wimplicit-int-float-conversion]
  size_t last_iteration_nrows = static_cast<size_t>(last_epoch * iteration_nrows);
                                                               ~ ^~~~~~~~~~~~~~~
src/core/models/dt_linearmodel.cc:619:52: note: in instantiation of function template specialization 'dt::LinearModel<double>::fit_impl<int>' requested here
template LinearModelFitOutput LinearModel<double>::fit_impl<int32_t>();
                                                   ^
src/core/models/dt_linearmodel.cc:131:68: warning: implicit conversion from 'size_t' (aka 'unsigned long') to 'double' may lose precision [-Wimplicit-int-float-conversion]
    iteration_nrows = static_cast<size_t>(std::ceil(nepochs_val_ * iteration_nrows));
                                                                 ~ ^~~~~~~~~~~~~~~
src/core/models/dt_linearmodel.cc:227:47: warning: implicit conversion from 'size_t' (aka 'unsigned long') to 'double' may lose precision [-Wimplicit-int-float-conversion]
                betas_[i][j] += betas[i][j] / dt::num_threads_in_team();
                                            ~ ^~~~~~~~~~~~~~~~~~~~~~~~~
src/core/models/dt_linearmodel.cc:297:74: warning: implicit conversion from 'size_t' (aka 'unsigned long') to 'double' may lose precision [-Wimplicit-int-float-conversion]
  double epoch_stopped = static_cast<double>(iteration_end) / dt_X_fit_->nrows();
                                                            ~ ~~~~~~~~~~~^~~~~~~
src/core/models/dt_linearmodel.cc:115:66: warning: implicit conversion from 'size_t' (aka 'unsigned long') to 'double' may lose precision [-Wimplicit-int-float-conversion]
  size_t last_iteration_nrows = static_cast<size_t>(last_epoch * iteration_nrows);
                                                               ~ ^~~~~~~~~~~~~~~
src/core/models/dt_linearmodel.cc:620:52: note: in instantiation of function template specialization 'dt::LinearModel<double>::fit_impl<double>' requested here
template LinearModelFitOutput LinearModel<double>::fit_impl<double>();
                                                   ^
src/core/models/dt_linearmodel.cc:131:68: warning: implicit conversion from 'size_t' (aka 'unsigned long') to 'double' may lose precision [-Wimplicit-int-float-conversion]
    iteration_nrows = static_cast<size_t>(std::ceil(nepochs_val_ * iteration_nrows));
                                                                 ~ ^~~~~~~~~~~~~~~
src/core/models/dt_linearmodel.cc:227:47: warning: implicit conversion from 'size_t' (aka 'unsigned long') to 'double' may lose precision [-Wimplicit-int-float-conversion]
                betas_[i][j] += betas[i][j] / dt::num_threads_in_team();
                                            ~ ^~~~~~~~~~~~~~~~~~~~~~~~~
src/core/models/dt_linearmodel.cc:297:74: warning: implicit conversion from 'size_t' (aka 'unsigned long') to 'double' may lose precision [-Wimplicit-int-float-conversion]
  double epoch_stopped = static_cast<double>(iteration_end) / dt_X_fit_->nrows();
                                                            ~ ~~~~~~~~~~~^~~~~~~
24 warnings generated.
```

--- a/src/core/models/dt_linearmodel.cc
+++ b/src/core/models/dt_linearmodel.cc
@@ -112,7 +112,9 @@ LinearModelFitOutput LinearModel<T>::fit_impl() {
   T last_epoch = nepochs_ - static_cast<T>(niterations) + 1;
 
   size_t iteration_nrows = dt_X_fit_->nrows();
-  size_t last_iteration_nrows = static_cast<size_t>(last_epoch * iteration_nrows);
+  // Note: if iteration_nrows > 16.78M, this conversion loses precision
+  T iteration_nrows_f = static_cast<T>(iteration_nrows);
+  size_t last_iteration_nrows = static_cast<size_t>(last_epoch * iteration_nrows_f);
   size_t total_nrows = (niterations - 1) * iteration_nrows + last_iteration_nrows;
   size_t iteration_end;
 
@@ -128,7 +130,7 @@ LinearModelFitOutput LinearModel<T>::fit_impl() {
 
   if (validation) {
     cols_val = make_casted_columns(dt_X_val_, stype_);
-    iteration_nrows = static_cast<size_t>(std::ceil(nepochs_val_ * iteration_nrows));
+    iteration_nrows = static_cast<size_t>(std::ceil(nepochs_val_ * iteration_nrows_f));
     niterations = total_nrows / iteration_nrows + (total_nrows % iteration_nrows > 0);
     loss_history.resize(val_niters_, 0.0);
   }
@@ -222,9 +224,10 @@ LinearModelFitOutput LinearModel<T>::fit_impl() {
           // Second, average the local coefficients
           {
             std::lock_guard<std::mutex> lock(m);
+            auto nth = static_cast<T>(dt::num_threads_in_team());
             for (size_t i = 0; i < dt_model->ncols(); ++i) {
               for (size_t j = 0; j < dt_model->nrows(); ++j) {
-                betas_[i][j] += betas[i][j] / dt::num_threads_in_team();
+                betas_[i][j] += betas[i][j] / nth;
               }
             }
           }
@@ -294,7 +297,7 @@ LinearModelFitOutput LinearModel<T>::fit_impl() {
   );
   job.done();
 
-  double epoch_stopped = static_cast<double>(iteration_end) / dt_X_fit_->nrows();
+  double epoch_stopped = static_cast<double>(iteration_end) / static_cast<double>(dt_X_fit_->nrows());
   LinearModelFitOutput res = {epoch_stopped, static_cast<double>(loss)};
 
   return res;

